File: ET3ALEM_SERVER/BusinessEntities/Models/QuizAttempt.cs
﻿using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;

namespace BusinessEntities.Models
{
    public class QuizAttempt
    {
        public int Id { get; set; }

        [Required] public string UserId { get; set; }

        public virtual User User { get; set; }

        [Required] public int QuizId { get; set; }

        public virtual Quiz Quiz { get; set; }
        public DateTime StartTime { get; set; }
        public List<QuestionAttempt> QuestionsAttempts { get; set; }
        public double Grade { get; set; }
        public bool IsGraded { get; set; }
        public bool IsSubmitted { get; set; }
        public DateTime? SubmitTime { get; set; }

        public void GradeQuiz()
        {
            Grade = 0;
            IsGraded = true;
            QuestionsAttempts.ForEach(qA =>
            {
                qA.GradeQuestion();
                Grade += qA.Grade;
                IsGraded &= qA.IsGraded;
            });
        }
    }
}

File: ET3ALEM_CLIENT_WEB/src/app/quiz/Model/quiz-attempt.ts
import { Transform, Type } from "class-transformer";
import { RegisterUser } from "src/app/auth/Model/User";
import { QuestionTypeResolver } from "src/app/question/shared/question-type-resolver";
import { Quiz } from "./quiz";
import {
  QuestionAttempt
} from "../../question/Models/questionAttempt/question-attempt";

export class QuizAttempt {
    Id: number;
    UserId: string;
    @Type(()=> RegisterUser)
    User;
    QuizId: number;
    @Type(() => Quiz)
    Quiz: Quiz;
    @Type(() => Date)
    StartTime: Date;
    @Type(() => QuestionAttempt)
    @Transform((value, object, type) => QuestionTypeResolver.getSpecificQuestionAttempt(value))
    QuestionsAttempts: Array<QuestionAttempt> = [];
    Grade: number;
    IsGraded: boolean;
    @Type(() => Date)
    SubmitTime?: Date;
    IsSubmitted: boolean;

    constructor(Id, UserId, QuizId, StartTime, QuestionAttempts = [], Grade = 0, IsGraded = false, SubmitTime?: Date) {
        this.Id = Id;
        this.UserId = UserId;
        this.QuizId = QuizId;
        this.StartTime = StartTime;
        this.QuestionsAttempts = QuestionAttempts;
        this.Grade = Grade;
        this.IsGraded = IsGraded;
        this.SubmitTime = SubmitTime;
    }

    UpdateQuestionTypes() {
        this.QuestionsAttempts?.forEach(qA => qA.UpdateQuestionType());
    }
}
